### DISCUSSION TITLE
Smooth transition from django-graphql-jwt to strawberry-django-auth

### DISCUSSION THREAD

**Discussion by dbartenstein** [nrbnlulu/strawberry-django-auth#392] [Q&A] [answered]
Hi all!

First of all congrats to the nice strawberry :strawberry: library and your great extension - it’s very intuitive to work with! :clap: 

We’re having a Django project with many GraphQL queries & mutations, all built using graphene-django. To be able to use it alongside strawberry, we are working on switching to graphene-django 3.x.

Our idea is to do the migration "step by step" and at first only introducing new queries / mutations using strawberry at the endpoint `/graphql2`, while keeping the existing graphene-django queries / mutations at `/graphql`. 
And yes, we’re also using [Django GraphQL JWT](https://django-graphql-jwt.domake.io/index.html).

Is it possible to continue using Django GraphQL JWT with all mutations on `/graphql` while taking advantage of the obtained JWT token on `/graphql2` for authenticated queries / mutations?

A first try ended with:
```
django_1     | SystemCheckError: System check identified some issues:
django_1     | 
django_1     | ERRORS:
django_1     | gqlauth.RefreshToken.user: (fields.E304) Reverse accessor for 'gqlauth.RefreshToken.user' clashes with reverse accessor for 'refresh_token.RefreshToken.user'.
django_1     |  HINT: Add or change a related_name argument to the definition for 'gqlauth.RefreshToken.user' or 'refresh_token.RefreshToken.user'.
django_1     | gqlauth.RefreshToken.user: (fields.E305) Reverse query name for 'gqlauth.RefreshToken.user' clashes with reverse query name for 'refresh_token.RefreshToken.user'.
django_1     |  HINT: Add or change a related_name argument to the definition for 'gqlauth.RefreshToken.user' or 'refresh_token.RefreshToken.user'.
django_1     | refresh_token.RefreshToken.user: (fields.E304) Reverse accessor for 'refresh_token.RefreshToken.user' clashes with reverse accessor for 'gqlauth.RefreshToken.user'.
django_1     |  HINT: Add or change a related_name argument to the definition for 'refresh_token.RefreshToken.user' or 'gqlauth.RefreshToken.user'.
django_1     | refresh_token.RefreshToken.user: (fields.E305) Reverse query name for 'refresh_token.RefreshToken.user' clashes with reverse query name for 'gqlauth.RefreshToken.user'.
django_1     |  HINT: Add or change a related_name argument to the definition for 'refresh_token.RefreshToken.user' or 'gqlauth.RefreshToken.user'.
```

Happy for any advice on how to best transition. Thanks!

**Comment by nrbnlulu** [comment]
Hey, well for starters you'll want to remove django-graphql-jwt from your installed apps (or rename that reverse accessor somehow).
Secondly you'll need to implement context provider for your graphene schema like this https://github.com/nrbnlulu/strawberry-django-auth/blob/64a27a5ccc161189786d88d7953fc54076fffc83/gqlauth/core/middlewares.py#L109

**Reply by dbartenstein** [reply]
> Hey, well for starters you'll want to remove django-graphql-jwt from your installed apps (or rename that reverse accessor somehow). Secondly you'll need to implement context provider for your graphene schema like this
> 
> https://github.com/nrbnlulu/strawberry-django-auth/blob/64a27a5ccc161189786d88d7953fc54076fffc83/gqlauth/core/middlewares.py#L109

@nrbnlulu thanks for your swift reply! Highly appreciated! :bow: 

My idea was to even continue using the JWT mutations of django-graphql-jwt for now and let the new queries / mutations developed with strawberry simply use the obtained JWT token (without using strawberry-django-auth’s JWT mutations for now).
Is this something that should be possible? Can the JWT token obtained via django-graphql-jwt simply be reused with strawberry-django-auth?

## Idea
### /graphql (graphene-django):
- JWT mutations + queries
- Existing mutations + queries

### /graphl2 (strawberry-django)
- New mutation + queries (use JWT auth obtained via /graphql)

In further increments re-implement existing mutations + queries for strawberry-django. The goal is reached when there’s nothing left in the `/graphql` (graphene-django) section.

**Reply by nrbnlulu** [reply]
Oh I see, I am not familiar with `django-graphql-jwt` though seems like you'll need to craft a solution like I posted above or use strawberry extensions... Unfortunately I think you wouldn't be able to use our mutations without using our models...

**Reply by dbartenstein** [reply]
> Oh I see, I am not familiar with `django-graphql-jwt` though seems like you'll need to craft a solution like I posted above or use strawberry extensions... Unfortunately I think you wouldn't be able to use our mutations without using our models...

Thanks @nrbnlulu! So using [strawberry extensions](https://strawberry.rocks/docs/guides/custom-extensions) for calling Django auth’s `login` when the user has a valid JWT token (obtained via django-graphql-jwt) could do the job, right?

**Reply by nrbnlulu** [reply, marked as answer]
I hope :upside_down_face:

**Reply by dbartenstein** [reply]
> I hope upside_down_face

@nrbnlulu thanks! I will keep you updated.

**Reply by dbartenstein** [reply]
@nrbnlulu 
I promised you an update. We have now moved forward.

### Phase 1
- Fix reverse accessor clashes by modifiying the related_name on django-graphql-jwt: https://github.com/zemtu/django-graphql-jwt/commit/c89f73c0c2bcb89b3c9929f8a892d3b91f93a387
- Introduce strawberry :strawberry: under a second endpoint `/graphql2`
- Implement JWTAuthenticationExtension for strawberry to allow JWT token authentication using the JWT token obtained via django-graphql-jwt

So far it looks promising. Next phases: use strawberry-django-auth mutations for obtaining JWT tokens.

**Reply by nrbnlulu** [reply]
nice, goodluck!
